api/osf_groups/views.py

```diff
@@ -0,0 +1,231 @@
+
+    def get_serializer_class(self):
+        if self.request.method == 'PUT' or self.request.method == 'PATCH' or self.request.method == 'DELETE':
+            return GroupMemberDetailSerializer
+        elif self.request.method == 'POST':
+            return GroupMemberCreateSerializer
+        else:
+            return GroupMemberSerializer
+
+    # overrides DestroyAPIView
+    def perform_destroy(self, instance):
+        group = self.get_osf_group()
+        auth = get_user_auth(self.request)
+        try:
+            group.remove_member(instance, auth)
+        except ValueError as e:
+            raise ValidationError(detail=str(e.message))
+
+
+class GroupMembersList(OSFGroupMemberBaseView, bulk_views.BulkUpdateJSONAPIView, bulk_views.BulkDestroyJSONAPIView, bulk_views.ListBulkCreateJSONAPIView, ListFilterMixin):
+    view_name = 'group-members'
+
+    # Overrides ListBulkCreateJSONAPIView
+    def get_queryset(self):
+        queryset = self.get_queryset_from_request()
+        if is_bulk_request(self.request):
+            user_ids = []
+            for user in self.request.data:
+                try:
+                    user_id = user['id'].split('-')[1]
+                except AttributeError:
+                    raise ValidationError('Member identifier not provided.')
+                except IndexError:
+                    raise ValidationError('Member identifier incorrectly formatted.')
+                else:
+                    user_ids.append(user_id)
+            queryset = queryset.filter(guids___id__in=user_ids)
+        return queryset
+
+    # Overrides ListFilterMixin
+    def get_default_queryset(self):
+        # Returns all members and managers of the OSF Group (User objects)
+        return self.get_osf_group().members
+
+    # Overrides ListBulkCreateJSONAPIView
+    def get_serializer_context(self):
+        context = super(GroupMembersList, self).get_serializer_context()
+        # Permissions check handled here - needed when performing write operations
+        context['group'] = self.get_osf_group()
+        return context
+
+    # Overrides BulkDestroyJSONAPIView
+    def get_requested_resources(self, request, request_data):
+        requested_ids = []
+        for data in request_data:
+            try:
+                requested_ids.append(data['id'].split('-')[1])
+            except IndexError:
+                raise ValidationError('Member identifier incorrectly formatted.')
+
+        resource_object_list = OSFUser.objects.filter(guids___id__in=requested_ids)
+        for resource in resource_object_list:
+            self._assert_member_belongs_to_group(resource)
+
+        if len(resource_object_list) != len(request_data):
+            raise ValidationError({'non_field_errors': 'Could not find all objects to delete.'})
+
+        return resource_object_list
+
+    # Overrides ListFilterMixin
+    def build_query_from_field(self, field_name, operation):
```

Comment:
I feel something like this is a bit cleaner 
```python
    def build_query_from_field(self, field_name, operation):
        
        if field_name != 'role':
            return super(GroupMembersList, self).build_query_from_field(field_name, operation)
        if operation['op'] != 'eq':
            raise InvalidFilterOperator(value=operation['op'], valid_operators=['eq'])
        
        role = operation['value'].lower().strip()
        if role not in GROUP_ROLES:
            raise InvalidFilterValue(value=operation['value'])
        else:
            group = self.get_osf_group(check_object_permissions=False)
            return Q(id__in=group.managers if role == MANAGER else group.members_only)
```

Reply:
Thanks, I borrowed some pieces from this.  I prefer the first thing to not be a negative, if A: do this, otherwise, do this.  Instead of if not A: move on.